### ISSUE TITLE
Huge discrepency between training results

### ISSUE THREAD

**Issue by sayakpaul** [apple/tensorflow_macos#119]
I am training a simple fully connected network on the FashionMNIST dataset and there's a huge discrepancy between the results I got from MLCompute and a standard Colab GPU. 

For MLCompute, I am using a MacBook Pro with the M1 chip (running macOS 11.0). Here's the code to reproduce the results -
[hyperparameter-tuning-tensorflow-scikit.zip](https://github.com/apple/tensorflow_macos/files/5768097/hyperparameter-tuning-tensorflow-scikit.zip)

Now, the results -

```
Epoch 1/25
938/938 [==============================] - 1s 797us/step - loss: 1.0739 - accuracy: 0.6448 - val_loss: 0.7662 - val_accuracy: 0.7241
Epoch 2/25
938/938 [==============================] - 0s 527us/step - loss: 0.8835 - accuracy: 0.7305 - val_loss: 0.8771 - val_accuracy: 0.7619
Epoch 3/25
938/938 [==============================] - 0s 524us/step - loss: 1.1470 - accuracy: 0.7229 - val_loss: 1.0002 - val_accuracy: 0.7474
Epoch 4/25
938/938 [==============================] - 0s 520us/step - loss: 1.3810 - accuracy: 0.7182 - val_loss: 1.0500 - val_accuracy: 0.7210
Epoch 5/25
938/938 [==============================] - 0s 520us/step - loss: 1.6114 - accuracy: 0.7181 - val_loss: 1.1036 - val_accuracy: 0.7670
...
Epoch 20/25
938/938 [==============================] - 0s 518us/step - loss: 33.5506 - accuracy: 0.7030 - val_loss: 20.7841 - val_accuracy: 0.7752
Epoch 21/25
938/938 [==============================] - 0s 520us/step - loss: 32.1645 - accuracy: 0.7030 - val_loss: 11.2156 - val_accuracy: 0.6959
Epoch 22/25
938/938 [==============================] - 0s 517us/step - loss: 42.0026 - accuracy: 0.6993 - val_loss: 85.1904 - val_accuracy: 0.6771
Epoch 23/25
938/938 [==============================] - 0s 522us/step - loss: 56.7750 - accuracy: 0.7011 - val_loss: 99.3805 - val_accuracy: 0.7257
Epoch 24/25
938/938 [==============================] - 0s 520us/step - loss: 75.6570 - accuracy: 0.6930 - val_loss: 45.5150 - val_accuracy: 0.7035
Epoch 25/25
938/938 [==============================] - 0s 517us/step - loss: 89.2049 - accuracy: 0.6954 - val_loss: 98.6771 - val_accuracy: 0.7598
[INFO] evaluating network...
              precision    recall  f1-score   support

         top       0.10      0.56      0.17      1000
     trouser       0.12      0.03      0.04      1000
    pullover       0.10      0.07      0.08      1000
       dress       0.10      0.02      0.03      1000
        coat       0.11      0.07      0.09      1000
      sandal       0.09      0.02      0.03      1000
       shirt       0.11      0.12      0.12      1000
     sneaker       0.12      0.03      0.05      1000
         bag       0.10      0.08      0.09      1000
  ankle boot       0.11      0.02      0.03      1000

    accuracy                           0.10     10000
   macro avg       0.11      0.10      0.07     10000
weighted avg       0.11      0.10      0.07     10000

When I am using the exact same network and training on Colab, the results are far better - 

```
Epoch 21/25
938/938 [==============================] - 2s 2ms/step - loss: 0.4867 - accuracy: 0.8300 - val_loss: 0.4497 - val_accuracy: 0.8458
Epoch 22/25
938/938 [==============================] - 2s 3ms/step - loss: 0.4755 - accuracy: 0.8344 - val_loss: 0.4544 - val_accuracy: 0.8522
Epoch 23/25
938/938 [==============================] - 2s 3ms/step - loss: 0.4900 - accuracy: 0.8256 - val_loss: 0.4488 - val_accuracy: 0.8551
Epoch 24/25
938/938 [==============================] - 2s 2ms/step - loss: 0.4895 - accuracy: 0.8323 - val_loss: 0.4834 - val_accuracy: 0.8447
Epoch 25/25
938/938 [==============================] - 2s 2ms/step - loss: 0.4738 - accuracy: 0.8310 - val_loss: 0.4590 - val_accuracy: 0.8478
[INFO] evaluating network...
              precision    recall  f1-score   support

         top       0.76      0.82      0.79      1000
     trouser       0.98      0.96      0.97      1000
    pullover       0.67      0.78      0.72      1000
       dress       0.82      0.89      0.85      1000
        coat       0.72      0.79      0.76      1000
      sandal       0.98      0.94      0.96      1000
       shirt       0.70      0.43      0.53      1000
     sneaker       0.90      0.97      0.93      1000
         bag       0.97      0.95      0.96      1000
  ankle boot       0.97      0.93      0.95      1000

    accuracy                           0.85     10000
   macro avg       0.85      0.85      0.84     10000
weighted avg       0.85      0.85      0.84     10000
```

Here's my Colab Notebook - https://colab.research.google.com/gist/sayakpaul/4f08a43a2c6e0e1c04607be038ca31df/tensorflow_keras_hyperparameter_tuning_scikit.ipynb?authuser=1. 

**Steps to run the Colab Notebook**

Upload the code zip I supplied above to Colab and unzip it. Inside the `train.py` script, uncomment the following lines - 

```
# Import mlcompute module to use the optional set_mlc_device API for device selection with ML Compute.
from tensorflow.python.compiler.mlcompute import mlcompute

# Select CPU device.
mlcompute.set_mlc_device(device_name="any") # Available options are 'cpu', 'gpu', and ‘any'.
```

Now, you are all set! 

What's happening here folks?

**Comment by ysimonx** [comment]
on my macbook air m1 with 16 Gb, i had

```
Epoch 1/10
1875/1875 [==============================] - 1s 354us/step - loss: 0.6305 - accuracy: 0.7806
Epoch 2/10
1875/1875 [==============================] - 1s 341us/step - loss: 0.3819 - accuracy: 0.8634
Epoch 3/10
1875/1875 [==============================] - 1s 369us/step - loss: 0.3424 - accuracy: 0.8749
Epoch 4/10
1875/1875 [==============================] - 1s 351us/step - loss: 0.3160 - accuracy: 0.8837
Epoch 5/10
1875/1875 [==============================] - 1s 342us/step - loss: 0.2970 - accuracy: 0.8913
Epoch 6/10
1875/1875 [==============================] - 1s 341us/step - loss: 0.2820 - accuracy: 0.8952
Epoch 7/10
1875/1875 [==============================] - 1s 341us/step - loss: 0.2656 - accuracy: 0.9005
Epoch 8/10
1875/1875 [==============================] - 1s 353us/step - loss: 0.2519 - accuracy: 0.9081
Epoch 9/10
1875/1875 [==============================] - 1s 353us/step - loss: 0.2443 - accuracy: 0.9099
Epoch 10/10
1875/1875 [==============================] - 1s 342us/step - loss: 0.2343 - accuracy: 0.9114
```

**Comment by hapasa** [comment]
Fyi, I've also found that all my own very simple experiments have failed to converge, while both Colab GPU training and my own CPU training give "good" results.

I haven't reported this, because the errors are so blatant I assumed those must be known.   So the training is fast, but results are useless is my current assessment of TensorFlow on M1 status.

**Comment by anna-tikhonova** [comment]
Thank you very much for reporting this. We will investigate and report back.

**Comment by koskimas** [comment]
Any news on this? I'm also unable to get simple convolutional networks to learn when the same code achieves good results on linux.

**Comment by sayakpaul** [comment]
@anna-tikhonova it's been a while. Possible to provide an update?

**Comment by koskimas** [comment]
I found the issue in my case. I'm using the `model.fit_generator` method and my generator is a class like this:

```
class ImageFileIterator(keras.preprocessing.image.Iterator):
  ...
  def _get_batches_of_transformed_samples(self, indexes):
    ...
    return x, y
```

If `y` (the expected output) was a python array I got horrible convergence. However if I simply wrap it to a numpy array `np.array(y)` I get the same results as with linux and tensorflow-gpu.

The python array "works" but it's somehow interpreted wrong by tensorflow_macos, whereas tensorflow-gpu interprets it correctly.
